File: apps/api/app/services/analysis.py
from collections import Counter, defaultdict
from datetime import date

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.base import CostFact, PriceFact, PriceList, Product, TheoreticalMargin
from app.domain.decimal_utils import decimal_to_str
from app.domain.effective import resolve_effective
from app.domain.simulation import simulate
from app.domain.types import Driver, EffectiveCandidate, SourceStatus
from app.schemas.api import AnalysisResponse, AnalysisRow, EffectiveView, PriceListView
from app.services.persistence import latest_batch


def _effective_view(value) -> EffectiveView:
    return EffectiveView(
        value=decimal_to_str(value.value),
        valid_from=value.valid_from,
        status=value.status,
        warnings=list(value.warnings),
    )


def list_price_lists(db: Session) -> list[PriceListView]:
    return [
        PriceListView(code=item.code, description=item.description)
        for item in db.scalars(select(PriceList).order_by(PriceList.code)).all()
    ]


def analyze(db: Session, query_date: date, price_list_code: str) -> AnalysisResponse:
    batch = latest_batch(db)
    if batch is None:
        raise ValueError("No committed import is available")
    price_list = db.scalar(select(PriceList).where(PriceList.code == price_list_code))
    if price_list is None:
        raise ValueError("Unknown price list")

    prices = db.scalars(
        select(PriceFact).where(
            PriceFact.batch_id == batch.id,
            PriceFact.price_list_code == price_list_code,
        )
    ).all()
    costs = db.scalars(select(CostFact).where(CostFact.batch_id == batch.id)).all()
    objectives = db.scalars(
        select(TheoreticalMargin).where(
            TheoreticalMargin.batch_id == batch.id,
            TheoreticalMargin.price_list_name == price_list.description,
            TheoreticalMargin.is_ambiguous.is_(False),
        )
    ).all()
    products = {item.code: item for item in db.scalars(select(Product)).all()}

    price_groups = defaultdict(list)
    for item in prices:
        price_groups[(item.branch_code, item.product_code)].append(item)
    cost_groups = defaultdict(list)
    for item in costs:
        cost_groups[(item.branch_code, item.product_code)].append(item)
    objective_map = {item.product_code: item.percentage for item in objectives}
    product_codes = {item.product_code for item in prices} | set(objective_map)

    rows: list[AnalysisRow] = []
    for product_code in sorted(product_codes):
        branches = {key[0] for key in price_groups if key[1] == product_code} or {"1"}
        for branch_code in sorted(branches):
            price_value = resolve_effective(
                [
                    EffectiveCandidate(
                        value=item.value,
                        valid_from=item.valid_from,
                        source_row=item.source_row,
                        batch_id=item.batch_id,
                        source_status=SourceStatus(item.source_status),
                    )
                    for item in price_groups[(branch_code, product_code)]
                ],
                query_date,
            )
            matching_costs = cost_groups.get((branch_code, product_code)) or [
                item for (branch, code), values in cost_groups.items() if code == product_code for item in values
            ]
            cost_value = resolve_effective(
                [
                    EffectiveCandidate(
                        value=item.value,
                        valid_from=item.valid_from,
                        source_row=item.source_row,
                        batch_id=item.batch_id,
                        source_status=SourceStatus(item.source_status),
                    )
                    for item in matching_costs
                ],
                query_date,
            )
            warnings = list(dict.fromkeys(price_value.warnings + cost_value.warnings))
            blocked = price_value.status == "conflict" or cost_value.status == "conflict"
            ideal_percent = objective_map.get(product_code)
            actual_gain_amount = None
            actual_gain_percent = None
            if not blocked and price_value.value is not None:
                actual = simulate(
                    cost=cost_value.value,
                    driver=Driver.PRICE,
                    driver_value=price_value.value,
                    ideal_percent=ideal_percent,
                    source_inactive="inactive_source" in warnings,
                    source_unknown="unknown_source_status" in warnings,
                )
                actual_gain_amount = decimal_to_str(actual.gain_amount)
                actual_gain_percent = decimal_to_str(actual.gain_percent)
                warnings.extend(actual.warnings)
            if ideal_percent is None:
                warnings.append("missing_ideal_margin")
            warnings = list(dict.fromkeys(warnings))
            data_status = "conflict" if blocked else "warning" if warnings else "ok"
            rows.append(
                AnalysisRow(
                    product_code=product_code,
                    description=products.get(product_code).description if products.get(product_code) else None,
                    branch_code=branch_code,
                    price_list_code=price_list.code,
                    price_list_name=price_list.description,
                    price=_effective_view(price_value),
                    cost=_effective_view(cost_value),
                    ideal_percent=decimal_to_str(ideal_percent),
                    actual_gain_amount=actual_gain_amount,
                    actual_gain_percent=actual_gain_percent,
                    data_status=data_status,
                    warnings=warnings,
                    simulation_blocked=blocked,
                )
            )

    counts = Counter(row.data_status for row in rows)
    return AnalysisResponse(
        query_date=query_date,
        price_list=PriceListView(code=price_list.code, description=price_list.description),
        rows=rows,
        counts={"total": len(rows), "ok": counts["ok"], "warning": counts["warning"], "conflict": counts["conflict"]},
    )


def product_history(db: Session, product_code: str, price_list_code: str) -> dict:
    batch = latest_batch(db)
    if batch is None:
        raise ValueError("No committed import is available")
    prices = db.scalars(
        select(PriceFact)
        .where(
            PriceFact.batch_id == batch.id,
            PriceFact.product_code == product_code,
            PriceFact.price_list_code == price_list_code,
        )
        .order_by(PriceFact.valid_from)
    ).all()
    costs = db.scalars(
        select(CostFact)
        .where(CostFact.batch_id == batch.id, CostFact.product_code == product_code)
        .order_by(CostFact.valid_from)
    ).all()
    return {
        "product_code": product_code,
        "price_list_code": price_list_code,
        "prices": [
            {"date": item.valid_from, "value": decimal_to_str(item.value), "source_row": item.source_row}
            for item in prices
        ],
        "costs": [
            {"date": item.valid_from, "value": decimal_to_str(item.value), "source_row": item.source_row}
            for item in costs
        ],
    }

File: apps/web/lib/domain/analysis.ts
// Análisis por lista y fecha: resuelve precio/costo vigentes, objetivo y estado
// de cada producto. Función pura (los datos se traen aparte) → autoridad de
// servidor y totalmente verificable. Port fiel de services/analysis.py.
import { Decimal, toStr } from "./decimal";
import { resolveEffective } from "./effective";
import { simulate } from "./simulation";
import type { EffectiveCandidate, EffectiveValue } from "./types";
import type { CostRecord, MarginRecord, PriceRecord } from "./importer";
import { NO_CAPABILITIES, type DataCapabilities } from "./capabilities";

export interface PriceListRef {
  code: string;
  description: string;
}

export interface ProductRef {
  code: string;
  description: string | null;
}

export interface EffectiveView {
  value: string | null;
  validFrom: string | null;
  status: string;
  warnings: string[];
}

export interface AnalysisRow {
  productCode: string;
  description: string | null;
  branchCode: string;
  priceListCode: string;
  priceListName: string;
  price: EffectiveView;
  cost: EffectiveView;
  idealPercent: string | null;
  actualGainAmount: string | null;
  actualGainPercent: string | null;
  dataStatus: "ok" | "warning" | "conflict";
  warnings: string[];
  simulationBlocked: boolean;
}

export interface AnalysisResponse {
  queryDate: string;
  priceList: PriceListRef;
  rows: AnalysisRow[];
  counts: { total: number; ok: number; warning: number; conflict: number };
  capabilities: DataCapabilities;
}

export interface AnalyzeInput {
  queryDate: string;
  priceList: PriceListRef;
  prices: PriceRecord[];
  costs: CostRecord[];
  margins: MarginRecord[];
  products: Map<string, ProductRef>;
  /** Capacidades de datos del lote; por defecto ninguna (volumen/rubro ausentes). */
  capabilities?: DataCapabilities;
}

function effView(value: EffectiveValue): EffectiveView {
  return {
    value: toStr(value.value),
    validFrom: value.validFrom,
    status: value.status,
    warnings: [...value.warnings],
  };
}

function toCandidate(item: { value: Decimal | null; validFrom: string; sourceRow: number; sourceStatus: EffectiveCandidate["sourceStatus"] }): EffectiveCandidate {
  return {
    value: item.value,
    validFrom: item.validFrom,
    sourceRow: item.sourceRow,
    sourceStatus: item.sourceStatus,
  };
}

export function analyze(input: AnalyzeInput): AnalysisResponse {
  const { queryDate, priceList, products } = input;

  const prices = input.prices.filter((p) => p.priceListCode === priceList.code);
  const costs = input.costs;
  const objectives = input.margins.filter(
    (m) => !m.isAmbiguous && m.priceListName === priceList.description,
  );

  const priceGroups = new Map<string, PriceRecord[]>();
  for (const item of prices) {
    const key = `${item.branchCode}|${item.productCode}`;
    (priceGroups.get(key) ?? priceGroups.set(key, []).get(key)!).push(item);
  }
  const costGroups = new Map<string, CostRecord[]>();
  for (const item of costs) {
    const key = `${item.branchCode}|${item.productCode}`;
    (costGroups.get(key) ?? costGroups.set(key, []).get(key)!).push(item);
  }

  const objectiveMap = new Map<string, Decimal | null>();
  for (const item of objectives) objectiveMap.set(item.productCode, item.percentage);

  const productCodes = new Set<string>([...prices.map((p) => p.productCode), ...objectiveMap.keys()]);

  const rows: AnalysisRow[] = [];
  for (const productCode of [...productCodes].sort()) {
    const branches = new Set<string>();
    for (const key of priceGroups.keys()) {
      const [branch, code] = key.split("|");
      if (code === productCode) branches.add(branch);
    }
    if (branches.size === 0) branches.add("1");

    for (const branchCode of [...branches].sort()) {
      const priceValue = resolveEffective(
        (priceGroups.get(`${branchCode}|${productCode}`) ?? []).map(toCandidate),
        queryDate,
      );

      let matchingCosts = costGroups.get(`${branchCode}|${productCode}`);
      if (!matchingCosts) {
        matchingCosts = costs.filter((c) => c.productCode === productCode);
      }
      const costValue = resolveEffective(matchingCosts.map(toCandidate), queryDate);

      const warnings = [...new Set([...priceValue.warnings, ...costValue.warnings])];
      const blocked = priceValue.status === "conflict" || costValue.status === "conflict";
      const idealPercent = objectiveMap.get(productCode) ?? null;

      let actualGainAmount: string | null = null;
      let actualGainPercent: string | null = null;
      if (!blocked && priceValue.value !== null) {
        const actual = simulate({
          cost: costValue.value,
          driver: "price",
          driverValue: priceValue.value,
          idealPercent,
          sourceInactive: warnings.includes("inactive_source"),
          sourceUnknown: warnings.includes("unknown_source_status"),
        });
        actualGainAmount = toStr(actual.gainAmount);
        actualGainPercent = toStr(actual.gainPercent);
        warnings.push(...actual.warnings);
      }
      if (idealPercent === null) warnings.push("missing_ideal_margin");
      const finalWarnings = [...new Set(warnings)];
      const dataStatus = blocked ? "conflict" : finalWarnings.length > 0 ? "warning" : "ok";

      rows.push({
        productCode,
        description: products.get(productCode)?.description ?? null,
        branchCode,
        priceListCode: priceList.code,
        priceListName: priceList.description,
        price: effView(priceValue),
        cost: effView(costValue),
        idealPercent: toStr(idealPercent),
        actualGainAmount,
        actualGainPercent,
        dataStatus,
        warnings: finalWarnings,
        simulationBlocked: blocked,
      });
    }
  }

  const counts = { total: rows.length, ok: 0, warning: 0, conflict: 0 };
  for (const row of rows) counts[row.dataStatus] += 1;

  return { queryDate, priceList, rows, counts, capabilities: input.capabilities ?? NO_CAPABILITIES };
}
